lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -2975,6 +2977,96 @@ SwiftLanguageRuntimeImpl::GetTypeRef(CompilerType type,
   return type_ref;
 }
 
+CompilerType SwiftLanguageRuntimeImpl::AdjustTypeForOriginallyDefinedInModule(
+    CompilerType type) {
+  auto ts = type.GetTypeSystem().dyn_cast_or_null<TypeSystemSwift>();
+  assert(ts);
+  if (!ts)
+    return type;
+
+  lldb::TypeSP lldb_type =
+      ts->GetTypeSystemSwiftTypeRef().GetCachedType(type.GetMangledTypeName());
+
+  if (!lldb_type) {
+    LLDB_LOGV(GetLog(LLDBLog::Types),
+              "[AdjustTypeForOriginallyDefinedInModule] Could not find lldb "
+              "type for type with mangled name {0}",
+              type.GetMangledTypeName());
+
+    return type;
+  }
+
+  std::vector<lldb_private::CompilerContext> declContext =
+      lldb_type->GetDeclContext();
+  if (declContext.empty() ||
+      declContext[0].kind != CompilerContextKind::Module) {
+    LLDB_LOG(GetLog(LLDBLog::Types),
+             "[AdjustTypeForOriginallyDefinedInModule] No module decl context "
+             "for type with mangled name {0}",
+             type.GetMangledTypeName());
+
+    return type;
+  }
+
+  auto module_context = declContext[0];
+
+  swift::Demangle::Demangler dem;
+  auto *node = dem.demangleSymbol(type.GetMangledTypeName());
+
+  assert(node && node->hasChildren());
+  if (!node || !node->hasChildren()) {
+    LLDB_LOG(GetLog(LLDBLog::Types),
+             "[AdjustTypeForOriginallyDefinedInModule] Unexpected node for "
+             "type with mangled name {0}",
+             type.GetMangledTypeName());
+
+    return type;
+  }
+
+  NodePointer parent = node;
+  NodePointer child = node->getFirstChild();
+
+  // Find the type's module node.
+  while (child && child->getKind() != Node::Kind::Module &&
+         child->hasChildren()) {
+    parent = child;
+    child = child->getFirstChild();
+  }
```

Comment:
what about a concrete generic type, would its type parameters need to have their modules replaced?